packages/loopring_v3/contracts/aux/transactions/TransactionReader.sol

```diff
@@ -15,119 +15,102 @@ import "../../core/impl/libtransactions/WithdrawTransaction.sol";
+        DepositTransaction.readTx(txData, 0, deposit);
     }
 
     function readWithdrawal(
         ExchangeData.Block memory _block,
-        uint txIdx
+        uint txIdx,
+        bytes memory txData
         )
         internal
         pure
-        returns (WithdrawTransaction.Withdrawal memory)
+        returns (WithdrawTransaction.Withdrawal memory withdrawal)
     {
-        bytes memory data = _block.readTx(txIdx, ExchangeData.TransactionType.WITHDRAWAL);
-        return WithdrawTransaction.readTx(data, 1);
+        _block.readTx(txIdx, txData);
+        WithdrawTransaction.readTx(txData, 0, withdrawal);
     }
 
     function readAmmUpdate(
         ExchangeData.Block memory _block,
-        uint txIdx
+        uint txIdx,
+        bytes memory txData
         )
         internal
         pure
-        returns (AmmUpdateTransaction.AmmUpdate memory)
+        returns (AmmUpdateTransaction.AmmUpdate memory ammUpdate)
     {
-        bytes memory data = _block.readTx(txIdx, ExchangeData.TransactionType.AMM_UPDATE);
-        return AmmUpdateTransaction.readTx(data, 1);
+        _block.readTx(txIdx, txData);
+        AmmUpdateTransaction.readTx(txData, 0, ammUpdate);
     }
 
     function readTransfer(
         ExchangeData.Block memory _block,
-        uint txIdx
+        uint txIdx,
+        bytes memory txData
         )
         internal
         pure
-        returns (TransferTransaction.Transfer memory)
+        returns (TransferTransaction.Transfer memory transfer)
     {
-        bytes memory data = _block.readTx(txIdx, ExchangeData.TransactionType.TRANSFER);
-        return TransferTransaction.readTx(data, 1);
+        _block.readTx(txIdx, txData);
+        TransferTransaction.readTx(txData, 0, transfer);
     }
 
     function readSignatureVerification(
         ExchangeData.Block memory _block,
-        uint txIdx
+        uint txIdx,
+        bytes memory txData
         )
         internal
         pure
-        returns (SignatureVerificationTransaction.SignatureVerification memory)
+        returns (SignatureVerificationTransaction.SignatureVerification memory verification)
     {
-        bytes memory data = _block.readTx(txIdx, ExchangeData.TransactionType.SIGNATURE_VERIFICATION);
-        return SignatureVerificationTransaction.readTx(data, 1);
+         _block.readTx(txIdx, txData);
+        SignatureVerificationTransaction.readTx(txData, 0, verification);
     }
 
     function readTx(
         ExchangeData.Block memory _block,
         uint txIdx,
-        ExchangeData.TransactionType txType
+        bytes memory txData
         )
         internal
         pure
-        returns (bytes memory data)
     {
-        data = _block.readTransactionData(txIdx);
-        require(txType == ExchangeData.TransactionType(data.toUint8(0)), "UNEXPTECTED_TX_TYPE");
+        _block.data.readTransactionData(txIdx, _block.blockSize, txData);
     }
 
-    function createMinimalBlock(
+    function extractTransactions(
```

Comment:
is `extractTransactions` the plural form of `readTx`? If we call `extractTransactions` with numTransactions === 1, then in `extractTransactions` we have `txData := add(txData, TX_DATA_AVAILABILITY_SIZE)` but in `readTx` we don't. Is this a bug?

Reply:
If `extractTransactions` is the plural form of `readTx`, maybe rename it to `readTxs`?

Reply:
> is `extractTransactions` the plural form of `readTx`? If we call `extractTransactions` with numTransactions === 1, then in `extractTransactions` we have `txData := add(txData, TX_DATA_AVAILABILITY_SIZE)` but in `readTx` we don't. Is this a bug?

It is indeed the plural. `txData` is just a local variable in the function and `TX_DATA_AVAILABILITY_SIZE` is added to jump ahead to the new position for the next loop iteration. So for numTransactions == 1 it's not needed, but it doesn't hurt.